core/src/test/java/org/elasticsearch/action/admin/cluster/reroute/ClusterRerouteRequestRoundTripTests.java

```diff
@@ -0,0 +1,162 @@
+import org.elasticsearch.cluster.routing.allocation.command.CancelAllocationCommand;
+import org.elasticsearch.cluster.routing.allocation.command.MoveAllocationCommand;
+import org.elasticsearch.common.ParseFieldMatcher;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.network.NetworkModule;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.admin.cluster.reroute.RestClusterRerouteAction;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.rest.FakeRestRequest;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static java.util.Collections.emptyMap;
+
+/**
+ * Test for serialization and parsing of {@link ClusterRerouteRequest} and its commands.
+ */
+public class ClusterRerouteRequestRoundTripTests extends ESTestCase {
+    private static final int ROUNDS = 30;
+    private final List<Supplier<AllocationCommand>> randomCommandGenerators = new ArrayList<>();
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private final AllocationCommandRegistry allocationCommandRegistry;
+
+    public ClusterRerouteRequestRoundTripTests() {
+        command(() -> new AllocateReplicaAllocationCommand(randomAsciiOfLengthBetween(2, 10), between(0, 1000),
+                randomAsciiOfLengthBetween(2, 10)));
+        command(() -> new AllocateEmptyPrimaryAllocationCommand(randomAsciiOfLengthBetween(2, 10), between(0, 1000),
+                randomAsciiOfLengthBetween(2, 10), randomBoolean()));
+        command(() -> new AllocateStalePrimaryAllocationCommand(randomAsciiOfLengthBetween(2, 10), between(0, 1000),
+                randomAsciiOfLengthBetween(2, 10), randomBoolean()));
+        command(() -> new CancelAllocationCommand(randomAsciiOfLengthBetween(2, 10), between(0, 1000),
+                randomAsciiOfLengthBetween(2, 10), randomBoolean()));
+        command(() -> new MoveAllocationCommand(randomAsciiOfLengthBetween(2, 10), between(0, 1000),
+                randomAsciiOfLengthBetween(2, 10), randomAsciiOfLengthBetween(2, 10)));
+        namedWriteableRegistry = new NamedWriteableRegistry();
+        allocationCommandRegistry = new NetworkModule(null, null, true, namedWriteableRegistry).getAllocationCommandRegistry();
+    }
+
+    /**
+     * Registers a random generator for a type of command.
+     */
+    protected void command(Supplier<AllocationCommand> generator) {
```

Comment:
do we have a plugin that does this? This seems complicated and over-engineered to me, if nobody uses this for real the reason why we have it will get lost.